CAS-370 CAS-367 Drawer layout and navigation bar

bychkovdmitry opened the pull request:
https://stream-io.atlassian.net/browse/CAS-367
https://stream-io.atlassian.net/browse/CAS-370

### Description

This PR is a starting point for the work on the new design tickets

- Implemented navigation drawer with current user info, actions to create new direct/group channel and a logout button
- Implemented bottom navigation with "Chats" and "Mentions" buttons with badges indicating unread count

### Checklist

- [X] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [X] PR targets the `develop` branch
- [ ] Changelog updated with client-facing changes
- [ ] New code is covered by unit tests
- [X] Comparison screenshots added for visual changes
- [X] Reviewers added

<img src="https://user-images.githubusercontent.com/9600921/97830025-bdf0f480-1cdc-11eb-8cc5-d84e2d1a7b9f.jpeg" width="200"/>                 <img src="https://user-images.githubusercontent.com/9600921/97830028-c1847b80-1cdc-11eb-981e-369d1b92a75a.jpeg" width="200"/>

adasiewiczr commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/common/Extensions.kt:
```diff
@@ -61,3 +73,14 @@ fun Fragment.initToolbar(toolbar: Toolbar) {
     }
 }
+
+fun BottomNavigationView.setBangeNumber(@IdRes menuItemId: Int, badgeNumber: Int) {
```
```suggestion
fun BottomNavigationView.setBadgeNumber(@IdRes menuItemId: Int, badgeNumber: Int) {
```

adasiewiczr commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/common/SingleLiveEvent.kt:
```diff
@@ -0,0 +1,45 @@
+class SingleLiveEvent<T> : MutableLiveData<T>() {
+
+    private val logger = ChatLogger.get("SingleLiveEvent")
+    private val mPending = AtomicBoolean(false)
```
I'm not a big Hungarian notation fan and I don't think we are using it in our code, so I would recommend to remove `m` prefix

adasiewiczr commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/home/HomeFragment.kt:
```diff
@@ -0,0 +1,95 @@
+        binding.bottomNavigationView.setBangeNumber(R.id.mentions_fragment, state.mentionsUnreadCount)
+
+        nameTextView.text = state.user.name
+        Glide.with(this)
```
I would recommend replacing Glide with Coil(https://github.com/coil-kt/coil). Coil is much faster comparing to Glide, it has user-friendly API and it's backed by Kotlin Coroutines. Moreover, we are currently replacing Glide with Coil (https://github.com/coil-kt/coil) in our SDK.

adasiewiczr commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/home/HomeFragmentViewModel.kt:
```diff
@@ -0,0 +1,61 @@
+        val user: User = User(),
+        val totalUnreadCount: Int = 0,
+        // TODO: implement unread mentions count
+        val mentionsUnreadCount: Int = 5
```
Let's add a Jira task for that

bychkovdmitry replied:
Done

adasiewiczr commented on stream-chat-android-ui-components-sample/src/main/res/layout/fragment_home.xml:
```diff
@@ -0,0 +1,65 @@
+            android:layout_marginBottom="8dp"
+            android:background="?selectableItemBackground"
+            android:drawableStart="@drawable/ic_sign_out"
+            android:drawablePadding="16dp"
```
What do you think about extracting common dimens to separate XML (e.g. 8, 16, 24, 48)?

adasiewiczr commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/home/HomeFragment.kt:
```diff
@@ -0,0 +1,95 @@
+import io.getstream.chat.android.client.models.name
+import io.getstream.chat.ui.sample.R
+import io.getstream.chat.ui.sample.common.navigateSafely
+import io.getstream.chat.ui.sample.common.setBangeNumber
```
```suggestion
import io.getstream.chat.ui.sample.common.setBadgeNumber
```

adasiewiczr commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/home/HomeFragment.kt:
```diff
@@ -0,0 +1,95 @@
+
+    private fun renderState(state: HomeFragmentViewModel.State) {
+        binding.bottomNavigationView.setBangeNumber(R.id.channels_fragment, state.totalUnreadCount)
+        binding.bottomNavigationView.setBangeNumber(R.id.mentions_fragment, state.mentionsUnreadCount)
```
```suggestion
        binding.bottomNavigationView.setBadgeNumber(R.id.mentions_fragment, state.mentionsUnreadCount)
```

adasiewiczr commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/home/HomeFragment.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    private fun renderState(state: HomeFragmentViewModel.State) {
+        binding.bottomNavigationView.setBangeNumber(R.id.channels_fragment, state.totalUnreadCount)
```
```suggestion
        binding.bottomNavigationView.setBadgeNumber(R.id.channels_fragment, state.totalUnreadCount)
```

adasiewiczr reviewed (changes requested):
Looks good, I've added small suggestions

samiuelson commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/home/HomeFragmentViewModel.kt:
```diff
@@ -0,0 +1,61 @@
+
+    init {
+        _state.value = State(
+            user = ChatClient.instance().getCurrentUser() ?: User(),
```
Let's give the `User()` some meaningful name. e.g. `val unauthorizedUser = User("some_unique_id?")` and reuse it also in the `State.user = unauthorizedUser`. Then it would be easy to check / compare against it in other layers, like:
```kotlin
someUser == HomeFragmentViewModel.unauthorizedUser
```

zsmb13 commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/common/SingleLiveEvent.kt:
```diff
@@ -0,0 +1,45 @@
+import io.getstream.chat.android.client.logger.ChatLogger
+import java.util.concurrent.atomic.AtomicBoolean
+
+class SingleLiveEvent<T> : MutableLiveData<T>() {
```
This common `SingleLiveEvent` implementation has significant issues, there are a lot of things it doesn't handle correctly. It can't have multiple observers, it may lose values if multiple ones are posted when its observer isn't active, and removing observers won't work as expected either. As a bonus, the syntax for sending events is woeful.

I've written about this problem [here](https://zsmb.co/thoughts-about-event-handling-on-android/) and gave a talk about how I implemented this in my own framework [here](https://youtu.be/4U7hxKAZ7qE?t=1382). Let's discuss it, and see what our needs for events are exactly, and which of the problems around them we want to solve.

zsmb13 commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/common/SingleLiveEvent.kt:
```diff
@@ -0,0 +1,45 @@
+    private val mPending = AtomicBoolean(false)
+
+    override fun observe(owner: LifecycleOwner, observer: Observer<in T>) {
+        if (hasActiveObservers()) {
+            logger.logD("Multiple observers registered but only one will be notified of changes.")
+        }
```
We should never have this fail silently like this. Instead, either support multiple observers, or explicitly crash if someone attempts to register a second observer. Debugging a connected observer that doesn't receive events is more than painful.

samiuelson commented on stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/mentions/MentionsFragment.kt:
```diff
@@ -0,0 +1,28 @@
+class MentionsFragment : Fragment() {
+
+    private var _binding: FragmentMentionsBinding? = null
+    private val binding get() = _binding!!
```
Do we really need to have both `_binding` and `binding` as both are private? For me, it's almost the same as having a single `lateinit var binding: FragmentMentionsBinding`.

zsmb13 replied:
We need the separate nullable `_binding` because we have to set it to `null` in `onDestroyView` to prevent leaks. You can't `null` out a `lateinit` variable.

bychkovdmitry replied:
With lateinit we won't be able to release binding in `onDestroyView()`:
```
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
```

Leaving as is for now as the solution with delegate is too complex for the sample app:
https://medium.com/@Zhuinden/simple-one-liner-viewbinding-in-fragments-and-activities-with-kotlin-961430c6c07c

bychkovdmitry replied:
Replaced glide with coil, introduced common spacing dimensions, introduced `unauthorizedUser` in HomeFragmentViewModel, fixed typo in `setBangeNumber` method

bychkovdmitry commented:
Replaced glide with coil, introduced common spacing dimensions, introduced unauthorizedUser in HomeFragmentViewModel, fixed typo in setBangeNumber method. For now leaving SingleLiveData, but modified it to crash on multiple observers.

adasiewiczr commented on stream-chat-android-ui-components-sample/src/main/res/layout/fragment_home.xml:
```diff
@@ -0,0 +1,65 @@
+            android:layout_width="match_parent"
+            android:layout_height="48dp"
+            android:layout_gravity="bottom"
+            android:layout_marginBottom="8dp"
```
Can be replaced with `spacing_small `

bychkovdmitry replied:
Done

adasiewiczr reviewed (approved):
LGTM

bychkovdmitry commented:
Removed SingleLiveData as we already had similar solution (https://github.com/GetStream/stream-chat-android/blob/master/stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/utils/LiveDataEvent.kt)
